### PR TITLE
Adding Sync implementation

### PR DESCRIPTION
**Pull request by prklm10**
`Adding Sync implementation`
- Adding 600s timeout in comparison call.

### CODE REVIEW

**Review by shantanuk-browserstack — DISMISSED** [review]

File: percy/lib/cli_wrapper.py [hunk 1]
```diff
@@ -42,13 +42,13 @@ def is_percy_enabled():
             log(e, on_debug=True)
             return False
 
-    def post_screenshots(self, name, tag, tiles, external_debug_url=None, ignored_elements_data=None, considered_elements_data=None):
-        body = self._request_body(name, tag, tiles, external_debug_url, ignored_elements_data, considered_elements_data)
+    def post_screenshots(self, name, tag, tiles, external_debug_url=None, ignored_elements_data=None, considered_elements_data=None, sync=False):
```
Comment (chinmay-browserstack) [inline percy/lib/cli_wrapper.py:45]: Don't pass default value for sync

File: percy/lib/cli_wrapper.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ def post_failed_event(self, error):
                 "errorKind": 'sdk'
             }
 
-            response = requests.post(f'{PERCY_CLI_API}/percy/events', json=body, timeout=30)
+            response = requests.post(f'{PERCY_CLI_API}/percy/events', json=body, timeout=600)
```
Comment (chinmay-browserstack) [inline percy/lib/cli_wrapper.py:68]: why increase timeout here?

File: percy/lib/cli_wrapper.py [hunk 3]
```diff
@@ -77,36 +77,38 @@ def post_failed_event(self, error):
             log(e, on_debug=True)
             return None
 
-    def post_poa_screenshots(self, name, session_id, command_executor_url, capabilities, desired_capabilities, options=None):
+    def post_poa_screenshots(self, name, session_id, command_executor_url, capabilities, desired_capabilities, options=None, sync=False):
```
Comment (chinmay-browserstack) [inline percy/lib/cli_wrapper.py:80]: remove default value for sync

File: percy/lib/percy_automate.py [hunk 4]
```diff
@@ -39,16 +40,18 @@ def screenshot(self, name: str, **kwargs):
 
             ignore_region_elements = [element.id for element in options.get(IGNORE_ELEMENT_KEY, [])]
             consider_region_elements = [element.id for element in options.get(CONSIDER_ELEMENT_KEY, [])]
+            sync = options.get(SYNC, False)
```
Comment (chinmay-browserstack) [inline percy/lib/percy_automate.py:43]: Remove default value

File: percy/providers/app_automate.py [hunk 5]
```diff
@@ -15,6 +15,7 @@ def supports(remote_url) -> bool:
 
     def screenshot(self, name: str, **kwargs):
         session_details = self.execute_percy_screenshot_begin(name)
+        sync = kwargs.get('sync', False)
```
Comment (chinmay-browserstack) [inline percy/providers/app_automate.py:18]: Same

File: percy/providers/app_automate.py [hunk 6]
```diff
@@ -24,9 +25,11 @@ def screenshot(self, name: str, **kwargs):
         try:
             response = super().screenshot(name, **kwargs)
             percy_screenshot_url = response.get('link', '')
-            self.execute_percy_screenshot_end(name, percy_screenshot_url, 'success')
+            self.execute_percy_screenshot_end(name, percy_screenshot_url, 'success', sync)
```
Comment (chinmay-browserstack) [inline percy/providers/app_automate.py:28]: why sync passed here?

File: percy/providers/app_automate.py [hunk 7]
```diff
@@ -90,15 +93,17 @@ def execute_percy_screenshot_begin(self, name):
...
         try:
             request_body = {
                 'action': 'percyScreenshot',
                 'arguments': {
                     'state': 'end',
                     'percyScreenshotUrl': percy_screenshot_url,
                     'name': name,
-                    'status': status }
+                    'status': status,
+                    'sync': sync
```
Comment (chinmay-browserstack) [inline percy/providers/app_automate.py:105]: Any reason to pass sync?

File: percy/lib/cli_wrapper.py (hunk 1 above)
Comment (prklm10) [reply percy/lib/cli_wrapper.py]: passed None

File: percy/lib/cli_wrapper.py (hunk 2 above)
Comment (prklm10) [reply percy/lib/cli_wrapper.py]: removed.

File: percy/lib/cli_wrapper.py (hunk 3 above)
Comment (prklm10) [reply percy/lib/cli_wrapper.py]: default to None

File: percy/lib/percy_automate.py (hunk 4 above)
Comment (prklm10) [reply percy/lib/percy_automate.py]: None

File: percy/providers/app_automate.py (hunk 5 above)
Comment (prklm10) [reply percy/providers/app_automate.py]: None

File: percy/providers/app_automate.py (hunk 6 above)
Comment (prklm10) [reply percy/providers/app_automate.py:28]: for instrumentation

File: percy/providers/app_automate.py (hunk 7 above)
Comment (prklm10) [reply percy/providers/app_automate.py:105]: for instrumentation.

**Review by Amit3200 — APPROVED** [review]

File: tests/test_screenshot.py [hunk 8]
```diff
@@ -287,5 +287,19 @@ def test_posts_screenshot_poa(self):
...
+
+        driver = Mock(spec=WebDriver)
+        driver.session_id = 'Dummy_session_id'
+        driver.capabilities = { 'key': 'value' }
+        driver.desired_capabilities = { 'key': 'value' }
+        driver.command_executor = Mock()
+        driver.command_executor._url = 'https://hub-cloud.browserstack.com/wd/hub'
+        self.mock_webdriver.capabilities = { 'key': 'value' }
+
+        self.assertEqual(percy_screenshot(driver, 'Snapshot 3', options = {'sync': True}), 'sync-data')
```
Comment (Amit3200) [inline tests/test_screenshot.py:302]: use variable for assertion matching not hardcoded value

File: tests/test_screenshot.py [hunk 9]
```diff
@@ -68,13 +68,13 @@ def mock_screenshot(fail=False):
         status=(500 if fail else 200),
     )
 
-def mock_poa_screenshot(fail=False):
+def mock_poa_screenshot(fail=False, sync=False):
+    body = '{ "success": ' + ("true" if not fail else 'false, "error": "test"') + "}"
+    if sync: body = '{ "success": "true", "data": "sync-data" }'
```
Comment (Amit3200) [inline tests/test_screenshot.py:73]: ideally this should be closer to actual data object.

**Review by Amit3200 — APPROVED** [review]

**Review by shantanuk-browserstack — APPROVED** [review]

**Review by ninadbstack — COMMENTED** [review]
blind approve as app percy owner

**Review by ninadbstack — APPROVED** [review]

**Pull request merged**